### PR TITLE
[mypyc]Add LLPritimiveOp, implement list.__getitem__ based on new LLPrimitiveOp

### PR DESCRIPTION
**Pull request by TH3CHARLie**
This is the first PR towards mypyc/mypyc#709.

In this PR, a new `RegisterOp` subclass `LLPrimitiveOp` is implemented. The new op is aiming to replace the current `PrimitiveOp` in the future. The major differences between them is that `LLPrimitiveOp` uses a list of ops as its actual description, which is similar to `BasicBlock` and is more abstract compared with using a C codegen callback function directly.

The top first priority of this PR is to present an initial design and therefore discuss and improve it. To demonstrate the idea, the `list.__getitem__` op is reimplemented using the new primitive op. The basic workflow is:

- During IR build, the `list.__getitem__` matches a `LLPrimitiveOp` registry and uses the callback to generate a list of ops that describe this particular op.
- During code generation, generating from this list of ops.

To this end, a new op `CFunctionCall` is also added to represent an external C function call in the IR. In the future, we may add more Ops/adjust based on the current IR, adding more expressiveness. 

This PR has some known bugs/hacks:

- The naming of registers. The current implementation uses a hack that any op of `LLPrimitiveOp`'s ops shares the same register name as `LLPrimitiveOp`. That's because in this PR we only consider a `LLPrimitiveOp` that can be implemented using a single `CFunctionCall`. Besides, the naming between the ops is also missing thus we cannot have more than 1 `CFunctionCall`. Fixing this would be of high priority once the prototype workflow is discussed and ready to go.

Again, the main focus of this PR would be presenting, discussing and modifying the general workflow and design for later works. Any comments are welcomed and highly appreciated!

### CODE REVIEW

**Comment by TH3CHARLie** [comment]
some weird error for macos build(definitely not the first time I saw it)

**Comment by ethanhs** [comment]
I believe Travis was doing maintenance today.

**Comment by JukkaL** [comment]
Thanks for the PR! I'll do at least a first review pass tomorrow.

**Comment by TH3CHARLie** [comment]
> Right now, primitive ops do double duty both as a mechanism for representing calls to C functions and as the mechanism for doing data driven selection of what primitives to call. Maybe instead of adding an `LLPrimitiveOp` as the mechanism for doing data driven IR generation, it would be better to have the data driven mechanism allow registering an IR generation callback directly, if we think that's an important case.

Great advice! I'll do the analysis and try to summarize something, and then report at the original issue to keep the discussion context complete.

**Comment by JukkaL** [comment]
@msullivan I basically agree with you. I was expecting that we'd special case the simple cases so that the definition of most primitives will not be any more complicated than what it is right now. My original idea was to have a generic callback-based thing that can be used for tricky cases, though I'm not sure if I explained it anywhere (at least well enough).

File: mypyc/ir/ops.py [hunk 1]
```diff
@@ -755,6 +759,85 @@ def accept(self, visitor: 'OpVisitor[T]') -> T:
         return visitor.visit_primitive_op(self)
 
 
+@trait
+class IRBuilderInterface:
```
[the remark this answers is not in the document]
Comment (JukkaL) [reply mypyc/ir/ops.py]: I'd prefer to avoid a circular dependency. One plausible-sounding option would be to have an abstract "empty" type here that we'd cast inside primitive op definitions (such as `int_ops.py`) to the concrete IR builder type using some helper function. I.e. here the type would only act as a tag and wouldn't contain any methods. The benefit would be that we can avoid circular dependencies and we don't need to declare all methods here. The drawbacks would be needing to use the cast helper in each callback, and losing a tiny bit of static type safety (which doesn't really matter, I think).

If most primitive op definitions are just simple calls etc., we can have special casing for them, which would allow us to avoid the extra boilerplate.

**Comment by TH3CHARLie** [comment]
> @msullivan I basically agree with you. I was expecting that we'd special case the simple cases so that the definition of most primitives will not be any more complicated than what it is right now. My original idea was to have a generic callback-based thing that can be used for tricky cases, though I'm not sure if I explained it anywhere (at least well enough).

Ummm, I think you mentioned the callback style in the original issue's discussion so it's definitely misunderstanding from my side. I thought that all primitives would use a callback-way to generate some more IRs and then generate C code from them, which is now presented in code.

So I am a little bit confused again now. My understanding:

1. For simple primitives (the ones we use `call_emit` now), the new design should have no more work than it does now: for example, the new primitive holds a `str` of the C function name and has args and ret_typ in its opdescrption(pretty much the same as now).

2. For complicated ones, we register multiple callbacks on it, each one corresponding to the different sets of primitives to be called on. For example, for `*` we have `list * int` or `int * list`, then we'll have two registered callbacks which generate two sets of IRs that handles processing the `int` and `list` and another simple primitive calling a C function, just like we described in 1.

Is understanding (partially) correct? @JukkaL @msullivan

**Comment by msullivan** [comment]
I think that understanding is correct, yeah.

I think there are a few possible options for how those callbacks are registered, exactly. The originally proposed approach is through LLPrimitiveOp, but it might also be worth considering decoupling data driven IR generation callbacks from primitive ops?

**Comment by JukkaL** [comment]
>  it might also be worth considering decoupling data driven IR generation callbacks from primitive ops?

After #8593, this now looks to me as the best option.

I think that this would mean that each primitive op could be one of these things only:
* Call a C function
* Get the address of a C name
* Read a C global

For calls to C functions, we can add some options to make this more general, such as these:

* Reordering arguments (for `PyDict_Contains`, at least)
* Specifying extra literal arguments, such as `NULL` or `-1` (useful for `str.split`, at least)
* Converting an int return value to `CPyTagged` (useful for `len(list)`, at least)

Alternatively, we can also define some additional small helper functions in `CPy.h`.

The key would be that all of these would be specified in data and it wouldn't be possible to attach an arbitrary callback to generate IR for a primitive op.

There would be some ops where the above wouldn't be enough, though there shouldn't be many of these. These would be special cased in `irbuild`. For some things we can use `irbuild.specialize`. We could generalize it to handle operators, but I think it would add unnecessary complexity. Instead, other things can be handled in an ad hoc way in`mypyc.irbuild.expression`, for example. We already do some special casing there, and adding some extra would be fine in my opinion.

**Comment by JukkaL** [comment]
If we drop IR generation callbacks from primitives, it looks like we wouldn't actually need `LLPrimitiveOp` -- we could directly generate C call ops (and address-of ops, etc.). We'd only need some form of low-level op description, and this could be defined in `mypyc.primitives.registry` (i.e. `mypyc.ir` wouldn't need to know about it). I'm starting to like this idea more and more.

**Comment by TH3CHARLie** [comment]
Thanks for the comments, I think I have something concrete to work with now, for example, implementing the `Call a C function` primitive.

However, there are two concepts that confuse me throughout the context:

- IR generation callbacks. Are we referring to callbacks that generate more IRs(like the style I presented in this PR), or are we referring to callbacks that generate C code from IR primitives(like the one used right now)?
- Data-driven. Is this referring to matching different types of args of the same primitive op(like we have four 'builtins.int's to handle four kinds of arg types)? And for this below:
> The key would be that all of these would be specified in data and it wouldn't be possible to attach an arbitrary callback to generate IR for a primitive op.
Does this mean, we no longer need a callback for C-function call primitive, and instead we build the primitive and fill its fields using only the information from the code itself(mypy AST)?
@JukkaL  @msullivan

**Comment by JukkaL** [comment]
> IR generation callbacks. Are we referring to callbacks that generate more IRs(like the style I presented in this PR), or are we referring to callbacks that generate C code from IR primitives(like the one used right now)?

The former. I.e., we'd no longer need to use any callback functions to define primitives after we've fully switched to the new approach. Any complex IR generation functionality would be moved to `mypyc.irbuild`. Instead of a callback, we'd just specify the name of C function being called (and any adjustments needed for arguments/return value) in the op description, in addition to the information we use to match the primitive op to AST.

> Data-driven. Is this referring to matching different types of args of the same primitive op(like we have four 'builtins.int's to handle four kinds of arg types)?

Oh I just meant that the IR output/semantics of primitives is also defined as data only, not via callbacks. Currently only the AST matching is data driven, but C generation (or IR generation in this PR) is still "code driven". Did this make things clearer?

> Does this mean, we no longer need a callback for C-function call primitive, and instead we build the primitive and fill its fields using only the information from the code itself(mypy AST)?

Basically yes, though we'd also use information from the op description (such as the name of C function to call), together with the mypy AST.

**Comment by TH3CHARLie** [comment]
Thanks for the elaboration!

I think I've got enough to start another try on the new approach(no more callback, build new primitives entirely in IR build with op descriptions). I'd like to close this PR since this is already pretty far from the new approach. A new PR would be nicer in terms of clearness and starting from #8593 would also prevent a lot of conflicts
